Update Package to Support PHP 8.0+ (v1.2)

justinyost opened the pull request:
Remove support for PHP 5/7. 

Updated CircleCI builds to install, test and code-sniff correctly on PHP 8.  

Updated all sub-classes for updated CodeSniffs, removes deprecation warnings for PHP 8, updated PHPUnit support to PHPUnit 9.0. 

Post this PR being merged, we're going to increment the version number to 1.2.

justinyost committed:
Test Multiple PHP Build Versions

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost committed:
Add Back a Build Job

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost committed:
Update to only Support PHP 8 and Above

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost committed:
Update CodeSniffer to v3

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost committed:
PHP Code Sniffer Improvements

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost committed:
Reduce Multiple Build Targets

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost committed:
Update PHPUnit Configuration for PHPUnit 9

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost committed:
Update PHPUnit to be PHPUnit 9

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost committed:
Ignore PHPUnit Results Cache

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost committed:
Run PHPUnit Tests in CircleCI

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost committed:
Update ArrayAccess Interface Method Definitions

Methods were sending deprecation warnings as we had issues on PHP 8

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost committed:
Update Test Class to Support PHPUnit 9

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost committed:
Update CircleCI Builds to Use Composer Scripts

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost committed:
Update PHPCS Configuration

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost committed:
PHPCBF Auto Applied Fixes for PSR-2

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost committed:
Corrected Namespace for DefaultApi Class

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost committed:
Updated Version Number to 1.2

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

lisavogtsf commented on src/com/amazon/paapi5/v1/BrowseNode.php:
```diff
@@ -401,7 +401,7 @@ public function setSalesRank($salesRank)
      * @return boolean
      */
-    public function offsetExists($offset)
+    public function offsetExists(mixed $offset): bool
```
Was this type hinting done as part of the code sniffing you mentioned?

justinyost replied:
Yep, we needed to add the typehints for the various [ArrayAccess](https://www.php.net/manual/en/class.arrayaccess.php) methods as they were presenting a Deprecation Warning before this.

lisavogtsf reviewed (commented):
Do we worry at all about staying in sync with Amazon? This feels like a lot of changes to something that is fundamentally Amazon's code.

Also is it expected that all the tests are 'Risky' or `This test did not perform any assertions` in local? Or am I possibly running them wrong?

enricobono commented on composer.json:
```diff
@@ -28,21 +28,21 @@
     ],
     "require": {
-        "php": ">=5.5",
+        "php": "^8.0",
```
Is this upgrade considered a _breaking change_ (thus we need to bump version to 2.0.0)?

justinyost replied:
@enricobono I don't think so based upon my understanding, referring to this post that sums up how I'm thinking about it https://medium.com/@sampart/semantic-versioning-when-you-change-the-required-programming-language-version-16a3a3555c95. 

The short version is: We're not changing the actual API for the package in any breaking way, merely the required PHP version and Composer will handle that if you don't have PHP 8 available on your machine you'll get the last version that does work for you.

enricobono replied:
Make sense, thanks.

justinyost commented:
@lisavogtsf

I don't necessarily want to be out of synch with Amazon's code, but also we need code that works for us and is easily installable. I'll include some documentation to note more explicitly that this package is not a 1:1 copy of Amazon's code in the Readme. 

Yes, that is indeed what the tests are reporting, I think it's a place we should look at making some improvements again beyond what Amazon have done, but not really in scope for our current work.

justinyost committed:
Updated Readme To List Changes from Amazon

Based upon feedback here: https://github.com/thewirecutter/paapi5-php-sdk/pull/13#pullrequestreview-1280061260

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>

justinyost commented:
@lisavogtsf Commit that adds explicit notes about what's changed from Amazon's original code: https://github.com/thewirecutter/paapi5-php-sdk/pull/13/commits/5ab2c8f87d71d7d927d9003c2b460a1f9c4de241

lisavogtsf reviewed (approved):
Thank you for the clarifications, this looks good to me, and I agree on the minor version change.

justinyost committed:
Remove Warning for Dynamic Property

Noted as an issue here: https://github.com/thewirecutter/paapi5-php-sdk/pull/12#issuecomment-1416219895

Signed-off-by: Justin Yost <justin.yost@wirecutter.com>
